### PR TITLE
Insights Operator pulling and updating pull-secret (transferring uncl…

### PR DESCRIPTION
**Pull request by tremes**
…assified clusters)

### CODE REVIEW

File: enhancements/insights/pulling-pull-secret-from-ocm.md
Comment (tremes) [inline enhancements/insights/pulling-pull-secret-from-ocm.md:155]: @mrunalp, @aravindhp What about these two chapters? Do we need to mention anything? WDYT pls?

Comment (aravindhp) [reply enhancements/insights/pulling-pull-secret-from-ocm.md:155]: I think we need to call out how we are going to handle scenarios where the cluster is not able to update the secret in the cluster during Window for whatever reason. Maybe send an event so that the user is aware?

File: enhancements/insights/pulling-pull-secret-from-ocm.md
Comment (wking) [inline enhancements/insights/pulling-pull-secret-from-ocm.md:69]: This makes me nervous.  The cluster's global pull secret is [explicitly doc'ed as admin-editable][1].  And removing pull-secret entries is [explicitly doc'ed as a way to disable remote monitoring][2].  This line will have the Insights operator and the local cluster-admin sharing responsibility for the global pull secret, and sharing is hard ;).  How will Insights distinguish entries that the local admin has appended?  Perhaps by assuming that RH-supplied pull secrets will never drop entries for a particular domain?  How will Insights distinguish entries that the local admin has explicitly removed?  Perhaps by using the OCM-API token alone as a marker for "there's a mismatch" and then only updating tokens for domains that exist in the current global pullspec?  Even if all that gets sorted, this is still granting the Insights operator huge power to unilaterally break the cluster (e.g. if someone manages to compromise or MitM the OCM-API and recommend a pivot to a garbage secret).  I think this needs a very detailed entry in the risks and mitigation section.

[1]: https://docs.openshift.com/container-platform/4.9/openshift_images/managing_images/using-image-pull-secrets.html#images-update-global-pull-secret_using-image-pull-secrets
[2]: https://docs.openshift.com/container-platform/4.9/support/remote_health_monitoring/opting-out-of-remote-health-reporting.html#insights-operator-new-pull-secret_opting-out-remote-health-reporting

File: enhancements/insights/pulling-pull-secret-from-ocm.md
Comment (wking) [inline enhancements/insights/pulling-pull-secret-from-ocm.md:70]: Which is great, unless the pools are paused ;).  If this lands, we'll need doc updates [here][1], and maybe in other places, that say "if you want your cluster to survive a cred handoff, be very sure to unpause and level all of your nodes before you disable the outgoing secret" or some such.

[1]: https://docs.openshift.com/container-platform/4.9/updating/update-using-custom-machine-config-pools.html#update-using-custom-machine-config-pools-about_update-using-custom-machine-config-pools

File: enhancements/insights/pulling-pull-secret-from-ocm.md
Comment (wking) [inline enhancements/insights/pulling-pull-secret-from-ocm.md:113]: And then what happens to this metric?  Is it covered by alerting?  Or exported via Telemetry/Insights?  Or...?

File: enhancements/insights/pulling-pull-secret-from-ocm.md
Comment (wking) [inline enhancements/insights/pulling-pull-secret-from-ocm.md:107]: nit: "tha the" -> "the"

File: enhancements/insights/pulling-pull-secret-from-ocm.md
Comment (wking) [inline enhancements/insights/pulling-pull-secret-from-ocm.md:142]: Assuming the in-cluster pull secret location doesn't move around, this seems safe.  But possibly worth talking through skew between different versions of the Insights operator and the OCM API service.  Or is the expectation that we get that API right on the first try and never have to update it going forward?

File: enhancements/insights/pulling-pull-secret-from-ocm.md
Comment (wking) [reply enhancements/insights/pulling-pull-secret-from-ocm.md:155]: Alerting, in both a public Insights rule and an on-cluster warning alert escalating to critical in the last day or something would be the loudest shouting for "the window is closing, didn't you want this handoff?!".  Might be worth giving folks a n OCM-side way to abort the handoff to silence the alarms if they'd changed their minds, although I guess OCM-UX is out-of-scope here.

File: enhancements/insights/pulling-pull-secret-from-ocm.md
Comment (wking) [inline enhancements/insights/pulling-pull-secret-from-ocm.md:163]: How often is Insights expected to poll the API?  If it's hourly or slower, it's hard to imagine this being a significant load on the OCM service side.  Are there latency constraints on the UX that make hourly polling for slow responses too slow?  Worth talking about latency constraints or lack thereof in the enhancement?

File: enhancements/insights/pulling-pull-secret-from-ocm.md
Comment (tremes) [reply enhancements/insights/pulling-pull-secret-from-ocm.md:113]: I think most likely alerting....and then maybe an Insights rule

File: enhancements/insights/pulling-pull-secret-from-ocm.md
Comment (tremes) [reply enhancements/insights/pulling-pull-secret-from-ocm.md:163]: Well we have been talking about daily interval AFAIK..... but @abhgupta can share his ideas/concerns

File: enhancements/insights/pulling-pull-secret-from-ocm.md
Comment (tremes) [reply enhancements/insights/pulling-pull-secret-from-ocm.md:155]: I think we won't be able to count the days within the window, but I think we can count the number of unsuccessful responses within the window and then fire an alert when a threshold is reached. WDYT?

File: enhancements/insights/pulling-pull-secret-from-ocm.md
Comment (tremes) [reply enhancements/insights/pulling-pull-secret-from-ocm.md:69]: My understanding is that the API will serve only the updated `pull-secret` (the update is a user action in the OCM IMO) and only within the given window. This should minimize the potential diffs between the secrets content. But yes, I agree that this should be described carefully. 
The only mitigation (in similar disaster scenarios) I can think of is to refer to the documentation and manual steps how to try again or update the secret manually. Are there any better ideas?

File: enhancements/insights/pulling-pull-secret-from-ocm.md
Comment (tremes) [reply enhancements/insights/pulling-pull-secret-from-ocm.md:142]: Yes I think we should say that the OCM API should preserve the backport compatibility and that the new version of the API will likely require an update of the Insights operator code. Would it be sufficient?

Comment (iNecas) [reply enhancements/insights/pulling-pull-secret-from-ocm.md:142]: I agree the probability of the API changing during the lifetime of the OCP version being supported should be very small and we should demand on the OCM API preserve the compatibility through the life-time of the versions depending on that particular API.

I can imagine having additional logic on specific extra status code form the API (let's say 410 Gone) and a follow-up alert from the insights-operator that it's outdated and the user needs to update for it to work properly: but personally, given its probability and the other ways there are in place notifying the user on using out-dated cluster, I would vote for simple logic and just assume the API will work on supported versions.

File: enhancements/insights/pulling-pull-secret-from-ocm.md
Comment (tremes) [reply enhancements/insights/pulling-pull-secret-from-ocm.md:155]: OK so the API will provide time information so we should be able to inform user about the last day or something...

File: enhancements/insights/pulling-pull-secret-from-ocm.md
Comment (tremes) [reply enhancements/insights/pulling-pull-secret-from-ocm.md:70]: OK makes sense. Do we need to update the enhancement to highlight this? @mrunalp can you help herer please.

File: enhancements/insights/pulling-pull-secret-from-ocm.md
Comment (tremes) [reply enhancements/insights/pulling-pull-secret-from-ocm.md]: I would resolve this conversation thread, but what about the `Failure Modes` chapter?

File: enhancements/insights/pulling-pull-secret-from-ocm.md
Comment (tremes) [reply enhancements/insights/pulling-pull-secret-from-ocm.md]: I updated this chapter. Please check.

File: enhancements/insights/pulling-pull-secret-from-ocm.md
Comment (aravindhp) [inline enhancements/insights/pulling-pull-secret-from-ocm.md:77]: You can link to the existing Jira epic / stories here

File: enhancements/insights/pulling-pull-secret-from-ocm.md
Comment (aravindhp) [inline enhancements/insights/pulling-pull-secret-from-ocm.md:106]: @mrunalp @markturansky should we add API details here once MR 2716 has merged?

File: enhancements/insights/pulling-pull-secret-from-ocm.md
Comment (aravindhp) [inline enhancements/insights/pulling-pull-secret-from-ocm.md:9]: Please @markturansky here as a reviewer.

File: enhancements/insights/pulling-pull-secret-from-ocm.md
Comment (aravindhp) [inline enhancements/insights/pulling-pull-secret-from-ocm.md:133]: Nit: `NA` might be more applicable than `Empty`

**Review by aravindhp — COMMENTED** [review]
Thanks for working on this @tremes

File: enhancements/insights/pulling-pull-secret-from-ocm.md
Comment (tremes) [reply enhancements/insights/pulling-pull-secret-from-ocm.md:77]: I will add the link to the `see also` section at the top

File: enhancements/insights/pulling-pull-secret-from-ocm.md
Comment (tremes) [reply enhancements/insights/pulling-pull-secret-from-ocm.md]: I mentioned the docs update requirement

File: enhancements/insights/pulling-pull-secret-from-ocm.md
Comment (wking) [reply enhancements/insights/pulling-pull-secret-from-ocm.md]: b55db2b012f9 looks good to me; thanks :)

File: enhancements/insights/pulling-pull-secret-from-ocm.md
Comment (wking) [reply enhancements/insights/pulling-pull-secret-from-ocm.md]: are we confident in the docs, or do we want a robotic guard in the Insights operator that blocks on and complains about paused pools?  Not clear from this doc is how long the outgoing pull-secret will continue to work.  If we're just passing a cluster between users, sticking with the outgoing secret for a while doesn't seem like a big deal.  If we're rotating because someone leaked the old secret, your paused-pool nodes might be break.  Also not clear to me is how gracefully nodes recover from stale pull secrets; is there anything in reboot/pivot code that could lock up?  Maybe all the in-cluster logic should live in the MCO and be about waving alarms if there's a paused pool and a pending pull-secret change?  Thoughts?

File: enhancements/insights/pulling-pull-secret-from-ocm.md
Comment (wking) [reply enhancements/insights/pulling-pull-secret-from-ocm.md:70]: is there a clear boundary for "the global pull secret provided by the OCM"?  There's a single global pull secret in-cluster, and the bits we hand out as an OCP pull secret could be some, all, or none of the cluster's current pull secret.  For example, maybe the cluster admin added a docker.io pull secret to the cluster's global secret.  We wouldn't want to clobber that.

File: enhancements/insights/pulling-pull-secret-from-ocm.md
Comment (wking) [reply enhancements/insights/pulling-pull-secret-from-ocm.md]: for Failure Modes we just dream up and discuss as many potential issues as we can think of:

* Some `paused` compute pool fails to pick up the new secret and continues attempting to use the old secret.  No problems if the old secret keeps working, but if it is revoked... FIXME: MCO/workloads folks might be able to clarify what will happen.
* Admin never confirms the handoff.  In this case, the window expires; no issues in-cluster.  If they open a support case "I thought I did transfer that over", support will `${STEPS}` to get logs showing that the transfer was initiated, alerted upon, but never confirmed in cluster. (possibly by visiting Telemetry, or maybe this would be summarized in an Insights rule?)
* Attacker who can forge an X.509 cert and impersonate OCM (or who gains server-side access to OCM) recommends an invalid pull secret.  Do we attempt to guard against this with some sort of pre-apply validation checks?  Do we mark it down as unlikely and have a generic "recovering your cluster from a broken pull secret" section coupled with our "remove `${KEY}` from your pull secret to tell Insights to stop polling upstream" section?
* More?

File: enhancements/insights/pulling-pull-secret-from-ocm.md
Comment (tremes) [reply enhancements/insights/pulling-pull-secret-from-ocm.md]: Thanks too.

File: enhancements/insights/pulling-pull-secret-from-ocm.md
Comment (tremes) [reply enhancements/insights/pulling-pull-secret-from-ocm.md:70]: @wking I added the following sentence there:

> Insights operator only replaces the entries provided in the updated `pull-secret` from the OCM API.

It means that the Insights operator cannot add or delete anything from the secret (I can try to emphasize this even more). It can only replace existing entries with new ones and if the entries (entry names) don't match then we should inform use (probably alert?)

File: enhancements/insights/pulling-pull-secret-from-ocm.md
Comment (tremes) [reply enhancements/insights/pulling-pull-secret-from-ocm.md]: Yep. Good points. Thanks.

- ad `#1` - We must mention the fact that the old secret will work for some time. Who can help us with the clarification (on the MCO side) please?
- ad `#2` - There's no such confirmation planned. We've discussed that the Insights operator is unable to provide such confirmation, because the "real" update is then up to the MCO. So we rely on Telemetry here (+ optionally an Insights rule)
- ad `#3` - Can we do some pre-apply validation @aravindhp? I don't think so, but it's definitely worth mentioning.

File: enhancements/insights/pulling-pull-secret-from-ocm.md
Comment (markturansky) [reply enhancements/insights/pulling-pull-secret-from-ocm.md:108]: @aravindhp  yes, and I think that MR can be wrapped up today. I'll be doing a final review with @abhgupta this afternoon.

File: enhancements/insights/pulling-pull-secret-from-ocm.md
Comment (markturansky) [reply enhancements/insights/pulling-pull-secret-from-ocm.md:173]: This API call will be for exactly 1 row in 1 table with a well-indexed ID field.  It returns an instance of "ClusterTransfer" with details populated or it's empty.  Either way, it should be very fast. This is a scalable solution.

File: enhancements/insights/pulling-pull-secret-from-ocm.md
Comment (wking) [reply enhancements/insights/pulling-pull-secret-from-ocm.md:70]: Probably also means that no operator besides Insights can use "remove an entry from the pull secret" as a knob for turning things off.

File: enhancements/insights/pulling-pull-secret-from-ocm.md
Comment (wking) [reply enhancements/insights/pulling-pull-secret-from-ocm.md]: I don't think you need to worry about MCO application for handoff confirmation.  The Insights operator knows and will alert on "pending transition waiting for local admin approval" per your existing proposal, and the Insights operator also knows if it accepts the approval and updates the local pull secret (regardless of whether the MCO then rolls that secret out).  So it should be possible to create an Insights rule that points out the most recent handoff attempt and then whether that handoff went through or not.  But I'm fine if the position is that, while a rule is possible, the risk of confusion is small enough that it's not worth writing the rule up, and we can always re-evaluate later if support bumps into this often enough that we want to revisit our guess at the risk size.

File: enhancements/insights/pulling-pull-secret-from-ocm.md
Comment (mrunalp) [reply enhancements/insights/pulling-pull-secret-from-ocm.md:108]: Yes, sounds good :+1:

File: enhancements/insights/pulling-pull-secret-from-ocm.md
Comment (iNecas) [reply enhancements/insights/pulling-pull-secret-from-ocm.md:70]: Yes, is there such a workflow today?

The one I know of is disabling the remote health functionality https://docs.openshift.com/container-platform/4.9/support/remote_health_monitoring/opting-out-of-remote-health-reporting.html.

If the remote health functionality is disabled, this workflow would not work at all, so implementaiton-wise probably not a worry, we should however make sure the docs are updated to ensure people realize this functionality would be lost as well.

Are there other cases where people might edit the relevant parts of pull secret outside of the knowledge of OCM?

**Comment by markturansky** [comment]
ClusterTransfer Spec was merged:  https://gitlab.cee.redhat.com/service/uhc-account-manager/-/merge_requests/2716

File: enhancements/insights/pulling-pull-secret-from-ocm.md
Comment (mrunalp) [inline enhancements/insights/pulling-pull-secret-from-ocm.md:54]: Customer should be able to trigger a pull secret update faster if they want to rather than having to wait for the full poll period (like 24 hours).

File: enhancements/insights/pulling-pull-secret-from-ocm.md
Comment (wking) [inline enhancements/insights/pulling-pull-secret-from-ocm.md:188]: While poking around in GitHub today, I stumbled across [ocm-agent][1].  Possibly worth talking to those folks about scoping and timing to see if you want to revisit this alternative or not.

[1]: https://github.com/openshift/ocm-agent

File: enhancements/insights/pulling-pull-secret-from-ocm.md
Comment (tremes) [reply enhancements/insights/pulling-pull-secret-from-ocm.md]: Yes. Thanks. I hopefully mentioned all the points here. Resolving the conversation.

File: enhancements/insights/pulling-pull-secret-from-ocm.md
Comment (tremes) [reply enhancements/insights/pulling-pull-secret-from-ocm.md:70]: We discussed this yesterday and I updated the enhancement.

File: enhancements/insights/pulling-pull-secret-from-ocm.md
Comment (tremes) [reply enhancements/insights/pulling-pull-secret-from-ocm.md:108]: I mentioned the `ClusterTransfer` API here and added the link to the MR.

File: enhancements/insights/pulling-pull-secret-from-ocm.md
Comment (tremes) [reply enhancements/insights/pulling-pull-secret-from-ocm.md]: I added one sentence describing briefly how to do this right now, but we will need to document it as well. I guess there will be a new documentation section/chapter related to this feature in the Insights operator.

File: enhancements/insights/pulling-pull-secret-from-ocm.md
Comment (tremes) [reply enhancements/insights/pulling-pull-secret-from-ocm.md]: I updated this section.

File: enhancements/insights/pulling-pull-secret-from-ocm.md
Comment (tremes) [reply enhancements/insights/pulling-pull-secret-from-ocm.md]: It's here https://github.com/openshift/enhancements/pull/971/files#diff-fb03640aacfca46208c1d4f791cc3fa8ad33e6644e236b9ddfac8a527296914cR54

File: enhancements/insights/pulling-pull-secret-from-ocm.md
Comment (aravindhp) [inline enhancements/insights/pulling-pull-secret-from-ocm.md:54]: >User can trigger the update faster by deleting the Insights operator Pod (This needs to be documented and may change in the future - e.g we can introduce a new CRD to configure this).

I suggest calling this out under `Risks and Mitigations`

File: enhancements/insights/pulling-pull-secret-from-ocm.md
Comment (aravindhp) [inline enhancements/insights/pulling-pull-secret-from-ocm.md:77]: I suggest making this a non-gaol.

File: enhancements/insights/pulling-pull-secret-from-ocm.md
Comment (aravindhp) [inline enhancements/insights/pulling-pull-secret-from-ocm.md:82]: You can also link to the stories here so that it is clearly called out:
```suggestion
- [CCXDEV-6529 Insights Operator pulling updated pull-secret to transfer unclassified cluster](https://issues.redhat.com/browse/CCXDEV-6529)
- [CCX-185 Transfer of Unclassified clusters - provide pull-secret](https://issues.redhat.com/browse/CCX-185)

File: enhancements/insights/pulling-pull-secret-from-ocm.md
Comment (aravindhp) [inline enhancements/insights/pulling-pull-secret-from-ocm.md:97]:
```suggestion
Risk: Users don't want to wait 24 hours for the pull secret to be updated on the cluster

Mitigation: User can trigger the update faster by deleting the Insights operator Pod (This needs to be documented and may change in the future - e.g we can introduce a new CRD to configure this).

```

File: enhancements/insights/pulling-pull-secret-from-ocm.md
Comment (aravindhp) [inline enhancements/insights/pulling-pull-secret-from-ocm.md:147]: This can also be N/A. The section is mainly for enhancements that involves removing a deprecated feature.

File: enhancements/insights/pulling-pull-secret-from-ocm.md
Comment (aravindhp) [reply enhancements/insights/pulling-pull-secret-from-ocm.md:187]: Thanks for pointing this out. The OCM folks are reviewing this enhancement, so they can comment. cc: @markturansky

**Review by aravindhp — CHANGES_REQUESTED** [review]
@tremes this looks really good. Just a few comments.

File: enhancements/insights/pulling-pull-secret-from-ocm.md
Comment (tremes) [reply enhancements/insights/pulling-pull-secret-from-ocm.md]: OK. Thanks. Done

File: enhancements/insights/pulling-pull-secret-from-ocm.md
Comment (tremes) [reply enhancements/insights/pulling-pull-secret-from-ocm.md]: Moved. Thanks

File: enhancements/insights/pulling-pull-secret-from-ocm.md
Comment (tremes) [reply enhancements/insights/pulling-pull-secret-from-ocm.md]: OK. Removed. Thank you.

File: enhancements/insights/pulling-pull-secret-from-ocm.md
Comment (markturansky) [reply enhancements/insights/pulling-pull-secret-from-ocm.md:190]: sounds good. future requirements TBD.

**Review by aravindhp — APPROVED** [review]
Thanks for working on this, @tremes 

/approve

@markturansky please review and LGTM

File: enhancements/insights/pulling-pull-secret-from-ocm.md
Comment (mrunalp) [inline enhancements/insights/pulling-pull-secret-from-ocm.md:138]: nit: typo `There`

File: enhancements/insights/pulling-pull-secret-from-ocm.md
Comment (mrunalp) [inline enhancements/insights/pulling-pull-secret-from-ocm.md:158]: typo: `Insights`

File: enhancements/insights/pulling-pull-secret-from-ocm.md
Comment (mrunalp) [inline enhancements/insights/pulling-pull-secret-from-ocm.md:174]: chapter -> section?

File: enhancements/insights/pulling-pull-secret-from-ocm.md
Comment (markturansky) [inline enhancements/insights/pulling-pull-secret-from-ocm.md:112]: you can change this link from MR to the doc itself:  https://gitlab.cee.redhat.com/service/uhc-account-manager/-/blob/master/docs/cluster_transfer_api.md

**Comment by markturansky** [comment]
LGTM but for one out-of-date link

File: enhancements/insights/pulling-pull-secret-from-ocm.md
Comment (tremes) [reply enhancements/insights/pulling-pull-secret-from-ocm.md]: Updated. Thanks

**Pull request merged**
